README.md
# Fontgen

Font rasterizer for Heaps based on Msdfgen. Primary target is automation of font rasterization with Signed Distance Field method.

## Usage
So far tool is built only with Hashlink VM and Windows, but with appropriate config should also work on Mac/Linux, as well as HXCPP/Eval/HLC.  
Command-line interface is currently limited to input config file and logger switches.  
Note that on Windows, HL should see the `msdfgen.dll` and `ammer_msdfgen.hdll`, so it should be either current working directory or in PATH.  
Warning: Some image viewers do not follow PNG specification properly and show some rasterized images as pure white.
This is an issue with particular decoder and file itself is correct. Known affected viewers: XNView, FastStone Image Viewer.

### General usage and config file structure.
Run tool with: `hl fontgen.hl <input file> [switches]`  
Input file should contain a JSON object or array of objects describing font rasterization parameters.
```haxe
{
	"input": String,
	"inputs": Array<String>,
	"output": String,
	"mode": SDFMode,
	"dfSize": Int,
	"charset": Array<CharsetConf>,
  "options": Array<ProcessorOption>,
	"fontSize": Int,
	"padding": { "top": Int, "bottom": Int, "left": Int, "right": Int },
  "spacing": { "x": Int, "y": Int },
  "packer": {
    "size": Int,
    "width": Int,
    "height": Int,
    "pot": Bool,
    "exact": Bool,
    "sort": String,
    "algorithm": String,
    "heuristic": String,
    "useWasteMap": String
  }
}
```
* `input` - path to TTF file relative to JSON file. Optional if `inputs` is present.
* `inputs` - An array of TTF files relative to JSON file. Can be used as fallback fonts in order of appearance. Optional if `input` is present.
* `output` - Path to output .fnt file relative to JSON file.
* `mode` - Describes mode in which glyphs are going to be generated. Possible values are: `msdf`, `sdf`, `psdf` and `raster`.
MSDF provides best accuracy by utilizing RGB channels, all others produce grayscale image.
* `dfSize` - Distance field range in pixels. Can be omitted for `raster` mode.
* `charset` - An array describing which characters should be rasterized in next priority:
  1. Try to find unicode block charset (or alias) described in `Charset.hx`
  2. If string points to a file, try to interpret it's contents.
    1. If file is a valid XML file, rasterizer will use characters present in all nodes that contain text. (Attributes and comments are ignored)
    2. Use contents of a file as a list of characters to rasterize.
  3. Use string as a list of characters to rasterize.
  
  Additionally, array nesting is supported.
* `fontSize` - desired font size in pixels.
* `padding` - optional and defaults to `0`. Describes extra padding for glyphs on the texture in pixels.
* `spacing` - optional and defaults to `1`. Describes spacing between glyphs on the texture in pixels.
* `options` - optional list of extra configuration flags. See below.

#### Packer configuration
Optional packer configuration allows to fine-tune atlas creation.
* `size` - Shortcut to set both `width` and `height` with same value. Default max atlas size is 4096x4096
* `width` - Maximum output atlas width in pixels. Note that it does not ensure atlas would be of specified size.
* `height` - Maximum output atlas height in pixels.
* `pot` - Output image will be power-of-two in dimensions. Note that this operation can override maximum size
(for example 122 max size can output 128px image), and packing of content happens within boundaries of define sizes, not power-of-two boundaries. Default: false
* `exact` - Ensures that output image is always of specified size. Default: false
* `sort` - Sets sorting function for glyphs prior packing. Prefixing name with `-` would reverse sort results.  
Available values: `width`, `height`, `area`, `perimeter` and `char`. Defaults to `-height`.  
* `algorithm` - Allows to select packing algorithm. Available values presented below. Defaults to `max-rect`
  * `guillotine` - Uses Guillotine algorithm. Currently not configurable and uses `BestLongSideFit` and `MaximizeArea` heuristics.
  * `naive-shelf` - Uses naive Shelf algorithm.
  * `shelf` - Uses Shelf algorithm. Default Heuristic is `BestArea`. Allows usage of waste map.
  * `simple-max-rect` - Uses simplified Max Rects algorithm.
  * `skyline` - Uses Skyline algorithm. Default Heuristic is `MinWasteFit`.
  * `max-rect` - Uses Max Rects algorithm. Default Heuristic is `BestLongSideFit`.
* `heuristic` - Specifies exact heuristic to use for some algorithms. **Case-sensitive.**
  * `shelf` - Dictates which shelf to choose:
    * `Next` - Always add the new rectangle to the last open shelf
    * `First` - Test each rectangle against each shelf in turn and pack it to the first where it fits
    * `BestArea` - Choose the shelf with smallest remaining shelf area
    * `WorstArea` - Choose the shelf with the largest remaining shelf area
    * `BestHeight` - Choose the smallest shelf (height-wise) where the rectangle fits
    * `BestWidth` - Choose the shelf that has the least remaining horizontal shelf space available after packing
    * `WorstWidth` - Choose the shelf that will have most remainining horizontal shelf space available after packing
  * `skyline` - Dictates prefered insertion location:
    * `BottomLeft`
    * `MinWasteFit`
  * `max-rect` - Dictates which rects is preferred when inserting:
    * `BestShortSideFit`
    * `BestLongSideFit`
    * `BestAreaFit`
    * `BottomLeftRule`
    * `ContactPointRule`
* `useWasteMap` - Enables or disables usage of waste maps for some algorithms.

See [`test/config.json`](test/config.json) for example config.

### Switches

* `-info` - enables logging of processing progress.
* `-nowarnings` - disables logging of processing warnings.
* `-timings` - enables logging of processing timings for each step.
* `-printmissing` - enables printing of all characters that were not present in ttf files but were required by charsets.
* `-silent` - suppresses all logging.
* `-verbose` - enables all logging.
* `-help` - Prints [help](src/help.txt) file and exits.

### Processor pptions
All processor options can be used as switches to enable them on all configurations. For example `-allownonprint`

* `allownonprint` - Enables rasterization of non-printing characters from range U+00 to U+1F plus U+7F (DEL). Disabled by default as to not produce warning about missing glyphs.
* `r8raster` - Forces greyscale rasterization without usage of alpha when using `raster` mode.

### .fnt file additions
Tool adds extra line at the end of `.fnt` file describing used SDF method that would allow decoder to determine parameters that are required to render the font.  
SDF descriptor presents itself as:
```
sdf mode=<mode> size=<dfSize>
```
It is put at the end of file, because some parsers depend on ordering of lines and may break if inserted between `info` and `chars` blocks.

## Compilation

TODO: More detailed
* Haxe Libraries:
  * [ammer](https://github.com/Aurel300/ammer/)
  * bin-packing (Haxelib version)
* Use MSVC in x86 mode, because ammer/hl don't like x64.
* Compile msdfgen in `native/msdfgen`  
Windows note: when using `cmake` it may fail at finding `freetype` lib. It's includes and .lib files are located in `native/msdfgen/freetype/` directory.
* Compile ammer library in `native`  
Mac/Linux note: Makefile is not valid, as it should also point at compiled `msdfgen.so/dylib` and `freetype.so/dylib`. Feel free to PR fixes, as I'm not using those OS ;)
* Run `build-hl.hxml` to compile hdll and the tool.  
Windows note: Make sure you are running in msvc envrionment, sicne ammer needs it to compile .hdll file.  
Ammer note: You probably will need to point at hashlink includes and lib files with `-D ammer.hl.hlInclude=<path-to-hashlink>/include -D ammer.hl.hlLibrary=<path-to-hashlink>`, see ammer lib for more details.
* Put `msdfgen.dll` from `native` and `ammer_msdfgen.hdll` from `native/hl` near `bin/fontgen.hl`
* You're good to go.

# License
* Source code is licensed under MIT
* `ammer`, `bin-packing` and `msdfgen` are under MIT
* Sample font files are licensed under SIL Open Font License, see [LICENSE](ttf/LICENSE/) folder for their respective license files.
* Msdfgen dependencies
  * `FreeType` - GPL2
  * `lodepng` - Zlib
  * `tinyxml` - Zlib

Welcome to licensing hell.

CHANGELOG.md
# 1.5.0
* Added packer configuration.
* Changed default max-rect heuristic to `BestLongSideFit` as it produce more compact results.
* Cahnged default sort to descending height (`-height`)
* Fixed graphical corruption when using padding with `raster` mode
* Mitigated graphical corruption on sdf due to normalization of a single dot.

# 1.4.0
* Added `everything` charset: Covers all unicode blocks.
* When using `everything` charset missing character counter is disabled
* When `charset` is not present in font config or is empty - `everything` is used.
* Added `LATIN1_SUPPLEMENT` to `LATIN` and `ANSI` aliases
* Added `ASCII` alias that covers `BASIC_LATIN` and `NONPRINTING`

# 1.3.0
* Added `r8raster` to use greyscale non-alpha image for rasterized glyphs.
* Fixed crash on same-directory config paths

# 1.2.0
* Added better solution for counter-clockwise glyphs.
* Removed `fixwinding` options (always-on from now on)
* Removed sdf chunk from .fnt file when in raster mode.

# 1.1.0
* Added config options feature
* Added `fixwinding` option to fix counter-clockwise glyphs.
* Added `allownonprinting` option to enable nonprinting range of characters
* Disabled nonprinting characters by default
* Added support for 1-bit (pure black and white) fonts in raster mode.

# 1.0.0
Initial release

native/msdfgen.cpp
#include <stdio.h>
#include <cstdlib>
#include <iostream>
#include "msdfgen_lib.h"
#include "msdfgen/msdfgen.h"
#include "msdfgen/msdfgen-ext.h"
#include <ft2build.h>
#include <lodepng.h>
#include FT_FREETYPE_H
#include FT_OUTLINE_H
#include FT_TRUETYPE_TABLES_H
#include FT_SFNT_NAMES_H

#ifdef __cplusplus
extern "C" {
#endif

using namespace msdfgen;

// Since we can't send structs/pointers to Haxe, store it locally and use slots for referencing.
struct FontSlot {
	FontHandle* font;
	FT_Face ft;
	double scale;
};

// Haxe struct for font metrics.
struct FontMetrics {
	int ascent;
	int descent;
	int unitsPerEm;
	int baseLine;
	int lineHeight;
	int flags;
};

// Haxe struct for glyph metrics.
struct GlyphMetrics {
	int width;
	int height;
	int offsetX;
	int offsetY;
	int advanceX;
	int descent;
};

FreetypeHandle* ft = NULL;
FT_Library ft_lib;
std::vector<FontSlot*> fonts;

int fontSize = 24;
double range = 4.0;
bool fixWinding = false;
Bitmap<byte, 4> atlasPixels;

LIB_EXPORT bool wrap_initializeFreetype() {
	ft = initializeFreetype();
	FT_Init_FreeType(&ft_lib);
	return ft != NULL;
}

LIB_EXPORT void wrap_deinitializeFreetype() {
	if ( ft != NULL ) {
		deinitializeFreetype(ft);
		FT_Done_FreeType(ft_lib);
	}
}

LIB_EXPORT void setParameters(double dfRange, int _fontSize, bool _fixWinding) {
	range = dfRange;
	fontSize = _fontSize;
	fixWinding = _fixWinding;
}

LIB_EXPORT int initFont(char* filename, unsigned char* metrics_data) {
	FontHandle* msdfHandle = loadFont(ft, filename);
	if (msdfHandle != NULL) {
		int index = fonts.size();
		struct FontSlot* slot = (FontSlot*)malloc(sizeof(FontSlot));
		slot->font = msdfHandle;
		
		FT_New_Face(ft_lib, filename, 0, &slot->ft);
		slot->scale = (double)fontSize / (double)slot->ft->units_per_EM * 64.;
		FT_Set_Pixel_Sizes(slot->ft, 0, fontSize);
		
		FontMetrics* metrics = reinterpret_cast<FontMetrics*>(metrics_data);
		metrics->ascent = slot->ft->ascender;
		metrics->descent = slot->ft->descender;
		metrics->unitsPerEm = slot->ft->units_per_EM;
		metrics->baseLine = (slot->ft->size->metrics.ascender + 32) >> 6;
		metrics->lineHeight = (slot->ft->size->metrics.height + 32) >> 6;
		TT_Header* header = (TT_Header*)FT_Get_Sfnt_Table(slot->ft, FT_SFNT_HEAD);
		metrics->flags = (int)header->Mac_Style | header->Flags << 16;
		
		fonts.push_back(slot);
		return index;
	}
	return -1;
}

// No single font unload, because there's no point.
LIB_EXPORT void unloadFonts() {
	std::vector<FontSlot*>::iterator it = fonts.begin();
	int size = fonts.size();
	while ( size-- > 0 ) {
		FontSlot* handle = *it++;
		destroyFont(handle->font);
		FT_Done_Face(handle->ft);
		free(handle);
	}
	fonts.clear();
}

LIB_EXPORT unsigned char* getFontName(int font, size_t *size) {
	FT_Face face = fonts[font]->ft;
	FT_SfntName name;
	int count = FT_Get_Sfnt_Name_Count(face);
	for (int i = 0; i < count; i++) {
		FT_Get_Sfnt_Name(face, i, &name);
		if (name.name_id == 4 && (name.platform_id == 3 || name.platform_id == 0) && name.language_id == 0x409) {
			unsigned char* data = (unsigned char*)malloc(name.string_len);
			*size = name.string_len;
			memcpy(data, name.string, name.string_len);
			return data;
		}
	}
	*size = 0;
	return nullptr;
}

LIB_EXPORT bool getGlyphMetrics(int font, int charcode, unsigned char* output) {
	FT_Face face = fonts[font]->ft;
	FT_UInt index = FT_Get_Char_Index(face, charcode);
	// In case we actually try to draw charcode 0.
	if (index == 0 && charcode != 0) return false;
	FT_Error err = FT_Load_Glyph(face, index, FT_LOAD_DEFAULT);
	if (err) return false;
	FT_GlyphSlot slot = face->glyph;
	GlyphMetrics* metrics = reinterpret_cast<GlyphMetrics*>(output);
	
	metrics->width = slot->bitmap.width;
	metrics->height = slot->bitmap.rows;
	metrics->offsetX = slot->bitmap_left;
	metrics->offsetY = slot->bitmap_top;
	metrics->advanceX = (slot->advance.x + 32) >> 6;
	metrics->descent = (slot->metrics.horiBearingY - slot->metrics.height) >> 6;
	// Slower alternative:
	// FT_BBox bbox;
	// FT_Outline_Get_CBox(&slot->outline, &bbox);
	// metrics->descent = bbox.yMin >> 6;
	return true;
}

LIB_EXPORT int getKerning(int font, int left, int right) {
	FT_Vector vec;
	FT_Face face = fonts[font]->ft;
	FT_Error err = FT_Get_Kerning(face, FT_Get_Char_Index(face, left), FT_Get_Char_Index(face, right), FT_KERNING_DEFAULT, &vec);
	if (err) return err;
	return (vec.x+32) >> 6;
}

LIB_EXPORT void beginAtlas(int atlasWidth, int atlasHeight, int defaultAlpha) {
	atlasPixels = Bitmap<byte, 4>(atlasWidth, atlasHeight);
	const int max = atlasWidth * atlasHeight * 4;
	unsigned char* pixels = (unsigned char*)atlasPixels;
	memset(pixels, 0, max); // Because for some reason there's garbage in allocated memory?
	if (defaultAlpha != 0) {
		char alpha = (char)defaultAlpha;
		for (int i = 3; i < max; i += 4) {
			pixels[i] = alpha;
		}
	}
}

LIB_EXPORT void endAtlas(char* output) {
	const unsigned char* pixels = (const unsigned char*)atlasPixels;
	lodepng::encode(output, pixels, atlasPixels.width(), atlasPixels.height(), LCT_RGBA);
	// TODO: Optimization: Save in appropriate format - grayscale (SDF/PSDF), RGB (MSDF) and RGBA for raster.
}

bool isInvalidWinding(Shape &shape) {
	double l, b, r, t;
	shape.bounds(l, b, r, t);
	Point2 p(l-(r-l)-1, b-(t-b)-1);
	double dummy;
	SignedDistance minDistance;
	for (std::vector<Contour>::const_iterator contour = shape.contours.begin(); contour != shape.contours.end(); ++contour) {
		for (std::vector<EdgeHolder>::const_iterator edge = contour->edges.begin(); edge != contour->edges.end(); ++edge) {
				SignedDistance distance = (*edge)->signedDistance(p, dummy);
				if (distance < minDistance)
						minDistance = distance;
		}
	}
	return minDistance.distance > 0;
}

LIB_EXPORT bool generateSDFGlyph(int slot, int charcode, int width, int height, int ox, int oy, double tx, double ty) {
	if (width == 0 || height == 0) return true;
	
	Shape glyph;
	if (loadGlyph(glyph, fonts[slot]->font, charcode)) {
		glyph.normalize();
		Bitmap<float, 1> sdf(width, height);
		double scale = fonts[slot]->scale;
		generateSDF(sdf, glyph, range / scale, scale, Vector2(tx/scale, ty/scale));
		oy += height;
		if (fixWinding && isInvalidWinding(glyph)) {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					byte px = pixelFloatToByte(1.f - *sdf(x, y));
					*it++ = px;
					*it++ = px;
					*it++ = px;
					*it++ = 0xff;
				}
			}
		} else {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					float px = pixelFloatToByte(*sdf(x, y));
					*it++ = px;
					*it++ = px;
					*it++ = px;
					*it++ = 0xff;
				}
			}
		}
		return true;
	}
	return false;
}

LIB_EXPORT bool generatePSDFGlyph(int slot, int charcode, int width, int height, int ox, int oy, double tx, double ty) {
	if (width == 0 || height == 0) return true;
	
	Shape glyph;
	if (loadGlyph(glyph, fonts[slot]->font, charcode)) {
		glyph.normalize();
		Bitmap<float, 1> sdf(width, height);
		double scale = fonts[slot]->scale;
		generatePseudoSDF(sdf, glyph, range / scale, scale, Vector2(tx/scale, ty/scale));
		oy += height;
		if (fixWinding && isInvalidWinding(glyph)) {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					byte px = pixelFloatToByte(1.f - *sdf(x, y));
					*it++ = px;
					*it++ = px;
					*it++ = px;
					*it++ = 0xff;
				}
			}
		} else {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					byte px = pixelFloatToByte(*sdf(x, y));
					*it++ = px;
					*it++ = px;
					*it++ = px;
					*it++ = 0xff;
				}
			}
		}
		return true;
	}
	return false;
}

LIB_EXPORT bool generateMSDFGlyph(int slot, int charcode, int width, int height, int ox, int oy, double tx, double ty) {
	if (width == 0 || height == 0) return true;
	Shape glyph;
	if (loadGlyph(glyph, fonts[slot]->font, charcode)) {
		glyph.normalize();
		edgeColoringSimple(glyph, 3, 0);
		Bitmap<float, 3> msdf(width, height);
		double scale = fonts[slot]->scale;
		generateMSDF(msdf, glyph, range / scale, scale, Vector2(tx/scale, ty/scale));
		oy += height;
		if (fixWinding && isInvalidWinding(glyph)) {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					*it++ = pixelFloatToByte(1.f - msdf(x, y)[0]);
					*it++ = pixelFloatToByte(1.f - msdf(x, y)[1]);
					*it++ = pixelFloatToByte(1.f - msdf(x, y)[2]);
					*it++ = 0xff;
				}
			}
		} else {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					*it++ = pixelFloatToByte(msdf(x, y)[0]);
					*it++ = pixelFloatToByte(msdf(x, y)[1]);
					*it++ = pixelFloatToByte(msdf(x, y)[2]);
					*it++ = 0xff;
				}
			}
		}
		return true;
	}
	return false;
}

LIB_EXPORT bool rasterizeGlyph(int slot, int charcode, int width, int height, int ox, int oy) {
	if (width == 0 || height == 0) return true;
	
	FT_Error err = FT_Load_Char(fonts[slot]->ft, charcode, FT_LOAD_RENDER);
	if (err) return false;
	FT_Bitmap* bitmap = &fonts[slot]->ft->glyph->bitmap;
	switch (bitmap->pixel_mode) {
		case FT_PIXEL_MODE_GRAY:
			for (int y = 0; y < height; y++) {
				byte* it = atlasPixels(ox, oy + y);
				for (int x = 0; x < width; x++) {
					float px = bitmap->buffer[(y) * bitmap->width + x];
					*it++ = 0xff;
					*it++ = 0xff;
					*it++ = 0xff;
					*it++ = px;
				}
			}
			
			return true;
			break;
		default:
			// TODO: Other pixel modes
			return false;
	}
}

#ifdef __cplusplus
}
#endif


native/msdfgen_lib.h
#pragma once

#ifdef __cplusplus
extern "C" {
#endif

#ifdef _WIN32
	#define LIB_EXPORT __declspec(dllexport)
#else
	#define LIB_EXPORT
#endif

LIB_EXPORT bool wrap_initializeFreetype();
LIB_EXPORT void wrap_deinitializeFreetype();

LIB_EXPORT void setParameters(double dfRange, int fontSize);
LIB_EXPORT int initFont(char* filename, unsigned char* metrics_data);
LIB_EXPORT void unloadFonts();
LIB_EXPORT bool getGlyphMetrics(int font, int charcode, unsigned char* output);
LIB_EXPORT int getKerning(int font, int left, int right);
LIB_EXPORT unsigned char* getFontName(int font, size_t* size);

LIB_EXPORT void beginAtlas(int atlasWidth, int atlasHeight, int defaultColor, bool _enforceR8);
LIB_EXPORT void endAtlas(char* output);

LIB_EXPORT bool generateSDFGlyph(int slot, int charcode, int width, int height, int ox, int oy, double tx, double ty, bool ccw);
LIB_EXPORT bool generatePSDFGlyph(int slot, int charcode, int width, int height, int ox, int oy, double tx, double ty, bool ccw);
LIB_EXPORT bool generateMSDFGlyph(int slot, int charcode, int width, int height, int ox, int oy, double tx, double ty, bool ccw);
LIB_EXPORT bool rasterizeGlyph(int slot, int charcode, int width, int height, int ox, int oy);

#ifdef __cplusplus
}
#endif

native/msdfgen_lib.cpp
#include <stdio.h>
#include <cstdlib>
#include <iostream>
#include "msdfgen_lib.h"
#include "msdfgen/msdfgen.h"
#include "msdfgen/msdfgen-ext.h"
#include <ft2build.h>
#include <lodepng.h>
#include FT_FREETYPE_H
#include FT_OUTLINE_H
#include FT_TRUETYPE_TABLES_H
#include FT_SFNT_NAMES_H
#include FT_BITMAP_H

#ifdef __cplusplus
extern "C" {
#endif

using namespace msdfgen;

// Since we can't send structs/pointers to Haxe, store it locally and use slots for referencing.
struct FontSlot {
	FontHandle* font;
	FT_Face ft;
	double scale;
};

// Haxe struct for font metrics.
struct FontMetrics {
	int ascent;
	int descent;
	int unitsPerEm;
	int baseLine;
	int lineHeight;
	int flags;
};

// Haxe struct for glyph metrics.
struct GlyphMetrics {
	int width;
	int height;
	int offsetX;
	int offsetY;
	int advanceX;
	int descent;
	int ccw;
};

FreetypeHandle* ft = NULL;
FT_Library ft_lib;
std::vector<FontSlot*> fonts;

int fontSize = 24;
bool enforceR8 = false;
bool normalizeShapes = false;
double range = 4.0;
Bitmap<byte, 4> atlasPixels;

LIB_EXPORT bool wrap_initializeFreetype() {
	ft = initializeFreetype();
	FT_Init_FreeType(&ft_lib);
	return ft != NULL;
}

LIB_EXPORT void wrap_deinitializeFreetype() {
	if ( ft != NULL ) {
		deinitializeFreetype(ft);
		FT_Done_FreeType(ft_lib);
	}
}

LIB_EXPORT void setParameters(double dfRange, int _fontSize) {
	range = dfRange;
	fontSize = _fontSize;
}

LIB_EXPORT int initFont(char* filename, unsigned char* metrics_data) {
	FontHandle* msdfHandle = loadFont(ft, filename);
	if (msdfHandle != NULL) {
		int index = fonts.size();
		struct FontSlot* slot = (FontSlot*)malloc(sizeof(FontSlot));
		slot->font = msdfHandle;
		
		FT_New_Face(ft_lib, filename, 0, &slot->ft);
		slot->scale = (double)fontSize / (double)slot->ft->units_per_EM * 64.;
		FT_Set_Pixel_Sizes(slot->ft, 0, fontSize);
		
		FontMetrics* metrics = reinterpret_cast<FontMetrics*>(metrics_data);
		metrics->ascent = slot->ft->ascender;
		metrics->descent = slot->ft->descender;
		metrics->unitsPerEm = slot->ft->units_per_EM;
		metrics->baseLine = (slot->ft->size->metrics.ascender + 32) >> 6;
		metrics->lineHeight = (slot->ft->size->metrics.height + 32) >> 6;
		TT_Header* header = (TT_Header*)FT_Get_Sfnt_Table(slot->ft, FT_SFNT_HEAD);
		metrics->flags = (int)header->Mac_Style | header->Flags << 16;
		
		fonts.push_back(slot);
		return index;
	}
	return -1;
}

// No single font unload, because there's no point.
LIB_EXPORT void unloadFonts() {
	std::vector<FontSlot*>::iterator it = fonts.begin();
	int size = fonts.size();
	while ( size-- > 0 ) {
		FontSlot* handle = *it++;
		destroyFont(handle->font);
		FT_Done_Face(handle->ft);
		free(handle);
	}
	fonts.clear();
}

LIB_EXPORT unsigned char* getFontName(int font, size_t *size) {
	FT_Face face = fonts[font]->ft;
	FT_SfntName name;
	int count = FT_Get_Sfnt_Name_Count(face);
	for (int i = 0; i < count; i++) {
		FT_Get_Sfnt_Name(face, i, &name);
		if (name.name_id == 4 && (name.platform_id == 3 || name.platform_id == 0) && name.language_id == 0x409) {
			unsigned char* data = (unsigned char*)malloc(name.string_len);
			*size = name.string_len;
			memcpy(data, name.string, name.string_len);
			return data;
		}
	}
	*size = 0;
	return nullptr;
}

LIB_EXPORT bool getGlyphMetrics(int font, int charcode, unsigned char* output) {
	FT_Face face = fonts[font]->ft;
	FT_UInt index = FT_Get_Char_Index(face, charcode);
	// In case we actually try to draw charcode 0.
	if (index == 0 && charcode != 0) return false;
	FT_Error err = FT_Load_Glyph(face, index, FT_LOAD_DEFAULT);
	if (err) return false;
	FT_GlyphSlot slot = face->glyph;
	GlyphMetrics* metrics = reinterpret_cast<GlyphMetrics*>(output);
	
	metrics->width = slot->bitmap.width;
	metrics->height = slot->bitmap.rows;
	metrics->offsetX = slot->bitmap_left;
	metrics->offsetY = slot->bitmap_top;
	metrics->advanceX = (slot->advance.x + 32) >> 6;
	metrics->descent = (slot->metrics.horiBearingY - slot->metrics.height) >> 6;
	metrics->ccw = FT_Outline_Get_Orientation(&slot->outline);
	// Slower alternative:
	// FT_BBox bbox;
	// FT_Outline_Get_CBox(&slot->outline, &bbox);
	// metrics->descent = bbox.yMin >> 6;
	return true;
}

LIB_EXPORT int getKerning(int font, int left, int right) {
	FT_Vector vec;
	FT_Face face = fonts[font]->ft;
	FT_Error err = FT_Get_Kerning(face, FT_Get_Char_Index(face, left), FT_Get_Char_Index(face, right), FT_KERNING_DEFAULT, &vec);
	if (err) return err;
	return (vec.x+32) >> 6;
}

LIB_EXPORT void beginAtlas(int atlasWidth, int atlasHeight, int defaultColor, bool _enforceR8) {
	atlasPixels = Bitmap<byte, 4>(atlasWidth, atlasHeight);
	enforceR8 = _enforceR8;
	const int max = atlasWidth * atlasHeight;
	unsigned int* pixels = (unsigned int*)(unsigned char*)atlasPixels;
	// Ensure empty memory
	memset(pixels, 0, max*sizeof(int));
	if (defaultColor != 0) {
		for (int i = 0; i < max; i++) {
			pixels[i] = defaultColor;
		}
	}
}

LIB_EXPORT void endAtlas(char* output) {
	const unsigned char* pixels = (const unsigned char*)atlasPixels;
	lodepng::encode(output, pixels, atlasPixels.width(), atlasPixels.height(), LCT_RGBA);
	// TODO: Optimization: Save in appropriate format - grayscale (SDF/PSDF), RGB (MSDF) and RGBA for raster.
}

void normalizeShape(Shape &shape) {
	if (normalizeShapes) {
		shape.normalize();
	} else {
		for (std::vector<msdfgen::Contour>::iterator contour = shape.contours.begin(); contour != shape.contours.end(); ++contour) {
			if (contour->edges.size() == 1) {
				contour->edges.clear();
			}
		}
	}
}

LIB_EXPORT bool generateSDFGlyph(int slot, int charcode, int width, int height, int ox, int oy, double tx, double ty, bool ccw) {
	if (width == 0 || height == 0) return true;
	
	Shape glyph;
	if (loadGlyph(glyph, fonts[slot]->font, charcode)) {
		normalizeShape(glyph);
		Bitmap<float, 1> sdf(width, height);
		double scale = fonts[slot]->scale;
		generateSDF(sdf, glyph, range / scale, scale, Vector2(tx/scale, ty/scale));
		oy += height;
		if (ccw) {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					byte px = pixelFloatToByte(1.f - *sdf(x, y));
					*it++ = px;
					*it++ = px;
					*it++ = px;
					*it++ = 0xff;
				}
			}
		} else {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					byte px = pixelFloatToByte(*sdf(x, y));
					*it++ = px;
					*it++ = px;
					*it++ = px;
					*it++ = 0xff;
				}
			}
		}
		return true;
	}
	return false;
}

LIB_EXPORT bool generatePSDFGlyph(int slot, int charcode, int width, int height, int ox, int oy, double tx, double ty, bool ccw) {
	if (width == 0 || height == 0) return true;
	
	Shape glyph;
	if (loadGlyph(glyph, fonts[slot]->font, charcode)) {
		normalizeShape(glyph);
		Bitmap<float, 1> sdf(width, height);
		double scale = fonts[slot]->scale;
		generatePseudoSDF(sdf, glyph, range / scale, scale, Vector2(tx/scale, ty/scale));
		oy += height;
		if (ccw) {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					byte px = pixelFloatToByte(1.f - *sdf(x, y));
					*it++ = px;
					*it++ = px;
					*it++ = px;
					*it++ = 0xff;
				}
			}
		} else {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					byte px = pixelFloatToByte(*sdf(x, y));
					*it++ = px;
					*it++ = px;
					*it++ = px;
					*it++ = 0xff;
				}
			}
		}
		return true;
	}
	return false;
}

LIB_EXPORT bool generateMSDFGlyph(int slot, int charcode, int width, int height, int ox, int oy, double tx, double ty, bool ccw) {
	if (width == 0 || height == 0) return true;
	Shape glyph;
	if (loadGlyph(glyph, fonts[slot]->font, charcode)) {
		normalizeShape(glyph);
		edgeColoringSimple(glyph, 3, 0);
		Bitmap<float, 3> msdf(width, height);
		double scale = fonts[slot]->scale;
		generateMSDF(msdf, glyph, range / scale, scale, Vector2(tx/scale, ty/scale));
		oy += height;
		if (ccw) {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					*it++ = pixelFloatToByte(1.f - msdf(x, y)[0]);
					*it++ = pixelFloatToByte(1.f - msdf(x, y)[1]);
					*it++ = pixelFloatToByte(1.f - msdf(x, y)[2]);
					*it++ = 0xff;
				}
			}
		} else {
			for (int y = height - 1; y >= 0; y--) {
				byte* it = atlasPixels(ox, oy - y);
				for (int x = 0; x < width; x++) {
					*it++ = pixelFloatToByte(msdf(x, y)[0]);
					*it++ = pixelFloatToByte(msdf(x, y)[1]);
					*it++ = pixelFloatToByte(msdf(x, y)[2]);
					*it++ = 0xff;
				}
			}
		}
		return true;
	}
	return false;
}

LIB_EXPORT bool rasterizeGlyph(int slot, int charcode, int width, int height, int ox, int oy) {
	if (width == 0 || height == 0) return true;
	
	FT_Error err = FT_Load_Char(fonts[slot]->ft, charcode, FT_LOAD_RENDER);
	if (err) return false;
	FT_Bitmap* bitmap = &fonts[slot]->ft->glyph->bitmap;
	int multiplier = 1;
	switch (bitmap->pixel_mode) {
		case FT_PIXEL_MODE_MONO:
			FT_Bitmap grayBtm;
			FT_Bitmap_Init(&grayBtm);
			FT_Bitmap_Convert(ft_lib, bitmap, &grayBtm, 1);
			bitmap = &grayBtm;
			multiplier = 0xff;
			// fall trough
		case FT_PIXEL_MODE_GRAY:
			if (enforceR8) {
				for (int y = 0; y < height; y++) {
					byte* it = atlasPixels(ox, oy + y);
					for (int x = 0; x < width; x++) {
						unsigned char px = bitmap->buffer[(y) * bitmap->width + x] * multiplier;
						*it++ = px;
						*it++ = px;
						*it++ = px;
						*it++ = 0xff;
					}
				}
			} else {
				for (int y = 0; y < height; y++) {
					byte* it = atlasPixels(ox, oy + y);
					for (int x = 0; x < width; x++) {
						unsigned char px = bitmap->buffer[(y) * bitmap->width + x] * multiplier;
						*it++ = 0xff;
						*it++ = 0xff;
						*it++ = 0xff;
						*it++ = px;
					}
				}
			}
			
			if (bitmap->pixel_mode == FT_PIXEL_MODE_MONO) {
				FT_Bitmap_Done(ft_lib, &grayBtm);
			}
			return true;
			break;
		default:
			std::cout << "[Error] Unsupported pixel mode: " << (int)bitmap->pixel_mode << "\n";
			// TODO: Other pixel modes
			return false;
	}
}

#ifdef __cplusplus
}
#endif
